src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,368 @@
+}
+
+bool QgsJoinByLocationAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  mBaseSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mBaseSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  mJoinSource.reset( parameterAsSource( parameters, QStringLiteral( "JOIN" ), context ) );
+  if ( !mJoinSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "JOIN" ) ) );
+
+
+  mJoinMethod = parameterAsEnum( parameters, QStringLiteral( "METHOD" ), context );
+
+  mPredicates = parameterAsEnums( parameters, QStringLiteral( "PREDICATE" ), context );
+
+  QString prefix = parameterAsString( parameters, QStringLiteral( "PREFIX" ), context );
+
+  if ( mFieldNames.empty() )
+  {
+    outFields2 = mJoinSource->fields();
+    fields2Indices.reserve( outFields2.count() );
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      fields2Indices << i;
+    }
+  }
+  else
+  {
+    fields2Indices.reserve( mFieldNames.count() );
+    for ( const QString &field : mFieldNames )
+    {
+      int index = mJoinSource->fields().lookupField( field );
+      if ( index >= 0 )
+      {
+        fields2Indices << index;
+        outFields2.append( mJoinSource->fields().at( index ) );
+      }
+    }
+  }
+
+  if ( !prefix.isEmpty() )
+  {
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      outFields2[ i ].setName( prefix + outFields2[ i ].name() );
+    }
+  }
+
+  if ( feedback->isCanceled() )
+    return false;
+
+  for ( int i = 0; i < fields2Indices.count(); ++i )
+    emptyAttrs << QVariant();
+
+  fields2Fetch = fields2Indices;
+  fields2Fetch << fields2Indices;
+
+  static const QgsFields mOutFields = QgsProcessingUtils::combineFields( mBaseSource->fields(), outFields2 );
+
+  joinedFeatures.reset( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, joinedFeaturesId, mOutFields,
+                                         mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+
+  if ( parameters.value( QStringLiteral( "OUTPUT" ) ).isValid() && !joinedFeatures )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  mDiscardNonMatching = parameterAsBoolean( parameters, QStringLiteral( "DISCARD_NONMATCHING" ), context );
+
+  if ( !mDiscardNonMatching )
+    {
+    unjoinedFeatures.reset( parameterAsSink( parameters, QStringLiteral( "NON_MATCHING" ), context, unjoinedFeaturesId, mBaseSource->fields(),
+                            mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+    if ( parameters.value( QStringLiteral( "NON_MATCHING" ) ).isValid() && !unjoinedFeatures )
+      throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "NON_MATCHING" ) ) );
+    }
```

Comment:
This code can be moved to processAlgorithm, it's all safe to do in a background thread

Reply:
Should be done